File: javascript/test/Behavioral/Observer.js
(function() {
  define(['Behavioral/Observer/DvdReleaseByCategory', 'Behavioral/Observer/DvdSubscriber', 'Behavioral/Observer/DvdRelease'], function(DvdReleaseByCategory, DvdSubscriber, DvdRelease) {
    'use strict';
    return describe('Test Behavioral Observer', function() {
      var btvs, btvsS2, jsopra, msimps, rgiles, simpS2, simpsons, smulde, soprS2, sopranos, wrosen, xfilS5, xfiles;
      btvs = new DvdReleaseByCategory("Buffy the Vampire Slayer");
      simpsons = new DvdReleaseByCategory("The Simpsons");
      sopranos = new DvdReleaseByCategory("The Sopranos");
      xfiles = new DvdReleaseByCategory("The X-Files");
      jsopra = new DvdSubscriber("Junior Soprano");
      msimps = new DvdSubscriber("Maggie Simpson");
      rgiles = new DvdSubscriber("Rupert Giles");
      smulde = new DvdSubscriber("Samantha Mulder");
      wrosen = new DvdSubscriber("Willow Rosenberg");
      btvsS2 = new DvdRelease("DVDFOXBTVSS20", "Buffy The Vampire Slayer Season 2", 2002, 6, 11);
      simpS2 = new DvdRelease("DVDFOXSIMPSO2", "The Simpsons Season 2", 2002, 7, 9);
      soprS2 = new DvdRelease("DVDHBOSOPRAS2", "The Sopranos Season 2", 2001, 11, 6);
      xfilS5 = new DvdRelease("DVDFOXXFILES5", "The X-Files Season 5", 2002, 4, 1);
      it('create a dvdreleasebycategory obj', function() {
        return expect(JSON.stringify(btvs)).toEqual('{"categoryName":"Buffy the Vampire Slayer","subscriberList":[],"dvdReleaseList":[]}');
      });
      it('create a dvdsubscriber obj', function() {
        return expect(JSON.stringify(jsopra)).toEqual('{"subscriberName":"Junior Soprano"}');
      });
      it('add dvdsubscribers', function() {
        expect(btvs.addSubscriber(rgiles)).toEqual(1);
        btvs.addSubscriber(wrosen);
        simpsons.addSubscriber(msimps);
        sopranos.addSubscriber(jsopra);
        xfiles.addSubscriber(smulde);
        return xfiles.addSubscriber(wrosen);
      });
      it('create a dvdrelease obj', function() {
        return expect(JSON.stringify(btvsS2)).toEqual('{"serialNumber":"DVDFOXBTVSS20","dvdName":"Buffy The Vampire Slayer Season 2","dvdReleaseYear":2002,"dvdReleaseMonth":6,"dvdReleaseDay":11}');
      });
      it('observe new dvdrelease', function() {
        expect(btvs.newDvdRelease(btvsS2).toString()).toEqual("Hello Rupert Giles , subscriber to the  Buffy the Vampire Slayer DVD release list.\nThe new Dvd Buffy The Vampire Slayer Season 2 will be released on 6/11/2002.,Hello Willow Rosenberg , subscriber to the  Buffy the Vampire Slayer DVD release list.\nThe new Dvd Buffy The Vampire Slayer Season 2 will be released on 6/11/2002.");
        simpsons.newDvdRelease(simpS2);
        sopranos.newDvdRelease(soprS2);
        return xfiles.newDvdRelease(xfilS5);
      });
      it('remove a subscriber', function() {
        return expect(xfiles.removeSubscriber(wrosen)).toBeTruthy();
      });
      it('update a release', function() {
        return expect(xfilS5.updateDvdReleaseDate(2002, 5, 14).toString()).toEqual('14');
      });
      return it('update a dvd', function() {
        return expect(xfiles.updateDvd(xfilS5).toString()).toEqual("Hello Samantha Mulder, subscriber to the The X-Files DVD release list.\nThe following DVDs release has been revised: The X-Files Season 5 will be released on 5/14/2002,Hello Willow Rosenberg, subscriber to the The X-Files DVD release list.\nThe following DVDs release has been revised: The X-Files Season 5 will be released on 5/14/2002");
      });
    });
  });

}).call(this);

File: scala/src/test/scala/Behavioral/Observer.scala
//Observer - testing the observer


package test.Behavioral;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.junit.Ignore;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;


import Behavioral.Observer.*;

public class Observer {

	@Test    
   public void obs() {
		DvdReleaseByCategory btvs = new DvdReleaseByCategory("Buffy the Vampire Slayer");
		DvdReleaseByCategory simpsons = new DvdReleaseByCategory("The Simpsons");  
		DvdReleaseByCategory sopranos = new DvdReleaseByCategory("The Sopranos");
		DvdReleaseByCategory xfiles = new DvdReleaseByCategory("The X-Files");

		DvdSubscriber jsopra = new DvdSubscriber("Junior Soprano");
		DvdSubscriber msimps = new DvdSubscriber("Maggie Simpson");
		DvdSubscriber rgiles = new DvdSubscriber("Rupert Giles");
		DvdSubscriber smulde = new DvdSubscriber("Samantha Mulder");
		DvdSubscriber wrosen = new DvdSubscriber("Willow Rosenberg");

		btvs.addSubscriber(rgiles);
		btvs.addSubscriber(wrosen);  
		simpsons.addSubscriber(msimps);
		sopranos.addSubscriber(jsopra);
		xfiles.addSubscriber(smulde);
		xfiles.addSubscriber(wrosen);

		DvdRelease btvsS2 = new DvdRelease("DVDFOXBTVSS20", "Buffy The Vampire Slayer Season 2", 2002, 06, 11);
		DvdRelease simpS2 = new DvdRelease("DVDFOXSIMPSO2", "The Simpsons Season 2", 2002, 07,  9);
		DvdRelease soprS2 = new DvdRelease("DVDHBOSOPRAS2", "The Sopranos Season 2", 2001, 11,  6);
		DvdRelease xfilS5 = new DvdRelease("DVDFOXXFILES5", "The X-Files Season 5", 2002, 04,  1);

		
		org.junit.Assert.assertEquals("New dvd release test 1", 	
			"Hello Rupert Giles, subscriber to the Buffy the Vampire Slayer DVD release list.\nThe new Dvd Buffy The Vampire Slayer Season 2 will be released on 6/11/2002.\nHello Willow Rosenberg, subscriber to the Buffy the Vampire Slayer DVD release list.\nThe new Dvd Buffy The Vampire Slayer Season 2 will be released on 6/11/2002.",
			btvs.newDvdRelease(btvsS2)
		);
		org.junit.Assert.assertEquals("New dvd release test 2", 	
			"Hello Maggie Simpson, subscriber to the The Simpsons DVD release list.\nThe new Dvd The Simpsons Season 2 will be released on 7/9/2002.",
			simpsons.newDvdRelease(simpS2)
		);
		org.junit.Assert.assertEquals("New dvd release test 3", 	
			"Hello Junior Soprano, subscriber to the The Sopranos DVD release list.\nThe new Dvd The Sopranos Season 2 will be released on 11/6/2001.",
			sopranos.newDvdRelease(soprS2)
		);
		org.junit.Assert.assertEquals("New dvd release test 4", 	
			"Hello Samantha Mulder, subscriber to the The X-Files DVD release list.\nThe new Dvd The X-Files Season 5 will be released on 4/1/2002.\nHello Willow Rosenberg, subscriber to the The X-Files DVD release list.\nThe new Dvd The X-Files Season 5 will be released on 4/1/2002.",
			xfiles.newDvdRelease(xfilS5)
		);

		org.junit.Assert.assertEquals("Remove subscriber test", 	
			true,
			xfiles.removeSubscriber(wrosen)
		);

		xfilS5.updateDvdReleaseDate(2002, 5, 14);

		org.junit.Assert.assertEquals("Update dvd test", 	
			"Hello Samantha Mulder, subscriber to the The X-Files DVD release list.\nThe following DVDs release has been revised: The X-Files Season 5 will be released on 5/14/2002.",
			xfiles.updateDvd(xfilS5)
		);

	}
}      
